pkg/domain/link.go

```diff
@@ -0,0 +1,224 @@
+			claim.ObjectMeta.Annotations = map[string]string{}
+		}
+		claim.ObjectMeta.Annotations[types.LastFailedAnnotationKey] = time.Now().Format(time.RFC3339)
+	}
+	claim.ObjectMeta.Annotations[types.StatusAnnotationKey] = text
+	err := c.updateFn(claim)
+	if err != nil {
+		c.logger("ClaimRedeemer", "Failed to update status for claim %q: %s", claim.ObjectMeta.Name, err)
+	}
+	if !failed {
+		return fmt.Errorf("Error processing claim %q: %s", claim.ObjectMeta.Name, text)
+	} else {
+		c.logger(c.name, "Failed to process claim %q: %s", claim.ObjectMeta.Name, text)
+		return nil
+	}
+}
+func (c *ClaimRedeemer) RedeemClaim(claim *corev1.Secret) error {
+
+	if claim.ObjectMeta.Annotations == nil {
+		return c.handleError(claim, "no annotations", true)
+	}
+	if claim.Data == nil {
+		return c.handleError(claim, "no data", true)
+	}
+	url, ok := claim.ObjectMeta.Annotations[types.ClaimUrlAnnotationKey]
+	if !ok {
+		return c.handleError(claim, "no url specified", true)
+	}
+	password, ok := claim.Data[types.ClaimPasswordDataKey]
+	if !ok {
+		return c.handleError(claim, "no password specified", true)
+	}
+	if failed, ok := claim.ObjectMeta.Annotations[types.LastFailedAnnotationKey]; ok {
+		c.logger(c.name, "Skipping failed claim %q (failed at %s)", claim.ObjectMeta.Name, failed)
+		return nil
+	}
+
+	ca, ok := claim.Data[types.ClaimCaCertDataKey]
+	transport := &http.Transport{}
+	if ok {
+		caPool := x509.NewCertPool()
+		caPool.AppendCertsFromPEM(ca)
+		transport.TLSClientConfig = &tls.Config{
+			RootCAs: caPool,
+		}
+	}
+	client := &http.Client{
+		Transport: transport,
+	}
+	request, err := http.NewRequest(http.MethodPost, url, bytes.NewReader(password))
+	request.Header.Add("skupper-site-name", c.siteId)
+	query := request.URL.Query()
+	query.Add("site-version", c.siteVersion)
+	request.URL.RawQuery = query.Encode()
+	response, err := client.Do(request)
+	if err != nil {
+		return c.handleError(claim, err.Error(), false)
+	}
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return c.handleError(claim, err.Error(), false)
+	}
+	if response.StatusCode != http.StatusOK {
+		fmt.Printf("Claim request failed with code: %d", response.StatusCode)
+		fmt.Println()
+		return c.handleError(claim, strings.TrimSpace(string(body)), response.StatusCode == http.StatusNotFound)
+	}
+	s := json.NewSerializerWithOptions(json.DefaultMetaFactory, scheme.Scheme, scheme.Scheme, json.SerializerOptions{Yaml: true})
+	var token corev1.Secret
+	_, _, err = s.Decode(body, nil, &token)
+	if err != nil {
+		return c.handleError(claim, "could not parse connection token", false)
+	}
+	for key, value := range token.ObjectMeta.Annotations {
+		claim.ObjectMeta.Annotations[key] = value
+	}
+	for key, value := range token.ObjectMeta.Labels {
+		claim.ObjectMeta.Labels[key] = value
+	}
+	claim.Data = token.Data
+	err = c.updateFn(claim)
+	if err != nil {
+		return fmt.Errorf("Could not store connection token for claim %q: %s", claim.ObjectMeta.Name, err)
```

Comment:
When trying to create link, I got the following:

Error: Could not store connection token for claim "link1": error creating token certificates for link link1 under volume skupper-router-certs - unable to create base directory link1-profile under volume skupper-router-certs - mkdir /home/username/.local/share/containers/storage/volumes/skupper-router-certs/_data/link1-profile: permission denied

Reply:
@ajssmith I have seen similar issues with volumes on Ubuntu, which seems to be related with user namespaces. I will test a few approaches to see if I can bypass it.

Reply:
@ajssmith it should have been fixed by: https://github.com/skupperproject/skupper/pull/908/commits/4461c0da130a6c1d5144b1fe4c3a820b6564e29f